### ISSUE TITLE
[BUG] Bad character in X-Spamd-Result: PHISHING

### ISSUE THREAD

**Issue by mguesdon** [rspamd/rspamd#3193]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:

**Describe the bug**

A control character in an URL can appear in the PHISHING part of X-Spamd-Result result. This make the header invalid and the mail is refused par software like Cyrus IMAP lmet.

### Steps to Reproduce

Write an mail with content like this:
--_000_AM0PR02MB38926A4A401FD310DC41B2F7D8540AM0PR02MB3892eurp_
Content-Type: text/plain; charset="iso-8859-1"
Content-Transfer-Encoding: quoted-printable

blabla
www.xxx.com<http://www.xxx.com%0d>
blabla

You'll see rspamd X-Spamd-Result: like this:
X-Spamd-Result: default: False [-7.09 / 20.00];
         PHISHING(4.00)[xxx.com->www.xxx.com^M];
         IP_SCORE(-3.70)[ip: (-9.73), ipnet: 8.3.5.0/24(-4.86), asn: 1234(-3.87), country: DE(-0.07)];

**Expected behavior**

^M should be removed before doing domain comparaison for phising
X-Spamd-Result: header should be filtered to remove illegal characters.

### Versions

rspamd 1.9.4
CentOS Linux release 7.5.1804

**Comment by vstakhov** [comment]
Unsupported version, closing.

**Comment by mguesdon** [comment]
OK. If someone use an old version and worry about this problem, here's a quick fix....


# diff /usr/share/rspamd/plugins/milter_headers.lua.1.9.4 /usr/share/rspamd/plugins/milter_headers.lua
243c243,246
<       if not s.options then s.options = {} end
---
>        if not s.options then s.options = {} end
>        -- MG-OXYMIUM: remove control characters like ^M
>        local s_options=table.concat(s.options, ',');
>        s_options = s_options:gsub('[%c]', '')
245c248
<         ' ', s.name, '(', string.format('%.2f', s.score), ')[', table.concat(s.options, ','), ']',
---
>         ' ', s.name, '(', string.format('%.2f', s.score), ')[', s_options, ']',

**Comment by Lekensteyn** [comment]
@mguesdon I was trying to see if this is still a problem with the current 2.2 release but your reproducer is damaged. Can you enclose it with triple backticks, like this:

   ```
   *formatting* here will be ignored
   ```

I came up with this, but this did not trigger the phishing score with `%0d` present:
```
From: <bob@example.com>
To: <alice@example.com>
Subject: test email
Message-ID: <test-1234@example.com>
Content-Type: text/html
Date: Wed, 15 Jan 2020 01:33:00 +0000

Test
<a href="http://example.com/">http://www.example.net%0d</a>
Bye
```

In any case, I believe that the problem was fixed in the 2.0 release via commit 9bd929050d737c61f0af5ae4c35faa181aecf20c.

**Comment by mguesdon** [comment]
@Lekensteyn 

Thank you. Here is a more complete fragment:
```
X-Greylist: Passed host: 78.153.245.27
Received: from smtpout21.mailinblack.com (smtpout21.mailinblack.com [78.153.245.27])
	by smtp-in2a-serv.oxymium.net (Postfix) with ESMTPS id B2C2B1C555C
	for <infos@yyyy.com>; Fri, 13 Dec 2019 16:22:49 +0100 (CET)
Received: from localhost (localhost [127.0.0.1])
	by smtpout21.mailinblack.com (Postfix) with ESMTP id 1D8CB147FD2
	for <infos@yyyy.com>; Fri, 13 Dec 2019 16:22:49 +0100 (CET)
Received: from mx21.mailinblack.com (localhost [127.0.0.1])
        by mib21.mailinblack.com
        with SMTP (Mib Daemon ) id K44AW2J6;
        Fri, 13 Dec 2019 16:22:48 +0100 (CET)
From: GNB <sup@xxxx.com>
To: A M <am@yyyy.com>, "infos@yyyy.com"
	<infos@yyyy.com>
CC: A B <ab@xxxx.com>, J M
	<jm@xxxx.com>, GNB <sup@xxxx.com>
Subject: RE: DISPONIBLE
Thread-Topic: DISPONIBLE
Thread-Index: AdA12Wl0amjoVw0gTb3RAmQzNX0JRB/0AAroF3wAT285cAAva0hYABl4gbQAHDhuUA=
Date: Fri, 13 Dec 2019 15:22:46 +0000
Message-ID:
 <AM0PR02MB38926A4A15401FD310DC41B2F7D8540@AM0PR02MB3892.eurprd02.prod.outlook.com>
References: <015801d55a5d1$a9d2cc90$fd7865b0$@yyyy.com>
 <AM0PR02MB38922FF87994521099C2C42BDBD8460@AM0PR02MB3892.eurprd02.prod.outlook.com>
 <AM0PR02MB3892766372A704574B98820CC4D85C0@AM0PR02MB3892.eurprd02.prod.outlook.com>
 <AM0PR02MB53781F90C9D4D3845553D1B3F2E6580@AM0PR02MB5378.eurprd02.prod.outlook.com>
 <AM0PR02MB3892523BF538387AD487B8AF17D85A0@AM0PR02MB3892.eurprd02.prod.outlook.com>
In-Reply-To:
 <AM0PR02MB3892523BF538387AD7B118AF17D85A0@AM0PR02MB3892.eurprd02.prod.outlook.com>
Accept-Language: fr-FR, en-US
Content-Language: fr-FR
Content-Type: multipart/related;
	boundary="_029_AM0PR02MB38926A4A401FD310DC41B2F7D8540AM0PR02MB3892eurp_";
	type="multipart/alternative"
MIME-Version: 1.0
X-Rspamd-Queue-Id: B2C2B1C555C
X-Spamd-Result: default: False [-7.09 / 20.00];
	 TO_DN_EQ_ADDR_SOME(0.00)[];
	 HAS_XOIP(0.00)[];
	 TO_DN_SOME(0.00)[];
	 RCPT_COUNT_FIVE(0.00)[5];
	 DKIM_TRACE(0.00)[zzzz.onmicrosoft.com:+];
	 MX_GOOD(-0.01)[xxxx-com02b.mail.protection.outlook.com,xxxx-com02b.mail.protection.outlook.com];
	 SUBJ_ALL_CAPS(1.43)[19];
	 FROM_EQ_ENVFROM(0.00)[];
	 MIME_TRACE(0.00)[0:+,1:+,2:+,3:~,4:~,5:~,6:~,7:~,8:~,9:~,10:~,11:~,12:~,13:~,14:~,15:~,16:~,17:~,18:~,19:~,20:~,21:~,22:~,23:~,24:~,25:~,26:~,27:~,28:~,29:~];
	 RCVD_TLS_LAST(0.00)[];
	 OXYMIUM_HOSTNAME_TLD(-10.00)[mailinblack.com];
	 ASN(0.00)[asn:12345, ipnet:1.2.3.0/24, country:DE];
	 PHISHING(4.00)[xxxx.com->www.xxxx.com
];
	 IP_SCORE(-3.70)[ip: (-9.73), ipnet: 78.153.245.0/24(-4.86), asn: 12345(-3.87), country: DE(-0.07)];
	 R_SPF_FAIL(1.00)[-all];
	 R_DKIM_ALLOW(-0.20)[zzzz.onmicrosoft.com:s=selector2-zzzz-onmicrosoft-com];
	 FROMDOMAIN_EQ_ENVFROMDOMAIN(0.00)[];
	 FROM_HAS_DN(0.00)[];
	 RCVD_COUNT_FIVE(0.00)[6];
	 MIME_GOOD(-0.10)[multipart/related,multipart/alternative,text/plain];
	 PREVIOUSLY_DELIVERED(0.00)[infos@yyyy.com];
	 MANY_INVISIBLE_PARTS(0.40)[5];
	 TO_MATCH_ENVRCPT_SOME(0.00)[];
	 NEURAL_HAM(-0.00)[-0.998,0];
	 OXYMIUM_SENDER_ALLOW(0.00)[Allowed: i-am-mx];
	 DMARC_POLICY_SOFTFAIL(0.10)[xxxx.com : No valid SPF, DKIM not aligned (relaxed),none]
X-Rspamd-Server: smtp-in2a-serv.oxymium.net
Deliver-To: Junk
X-Spam-Flag: YES

--_029_AM0PR02MB38926A4A401FD310DC41B2F7D8540AM0PR02MB3892eurp_
Content-Type: multipart/alternative;
	boundary="_000_AM0PR02MB38926A4A401FD310DC41B2F7D8540AM0PR02MB3892eurp_"

--_000_AM0PR02MB38926A4A401FD310DC41B2F7D8540AM0PR02MB3892eurp_
Content-Type: text/plain; charset="iso-8859-1"
Content-Transfer-Encoding: quoted-printable

Bonjour,
A =E0 B C D =
de l=
command=E9s.
donner ?

Cordialement

[cid:image001.jpg@01D5B1D1.8B2E]
sup@xxxx.com
[FB]<https://www.facebook.com/zzzz>[TWI]<https://twitter.c=
om/aaaa>[LINK]<https://www.linkedin.com/company/bbbb/?=
trk=3Dbiz>[INSTA]<https://www.instagram.com/zzzz/>[Y=
OU]<https://www.youtube.com/channel/UC3>
www.xxxx.com<http://www.xxxx.com%0d>

[Une image contenant clipart  Description g=E9n=E9r=E9e automatiquement]


--_000_AM0PR02MB38926A4A401FD310DC41B2F7D8540AM0PR02MB3892eurp_--

--_029_AM0PR02MB38926A4A401FD310DC41B2F7D8540AM0PR02MB3892eurp_--
```

It come from a "postcat" command on the message at the cyrus-imap mail server, edited to remove private informations.
I think you can remove X-Spam-* and feed a postfix server with it... PHISHING was triggered because "www.xxxx.com%0d" is not equal to "www.xxxx.com"
